src/csharp/Microsoft.Spark/Sql/CastUnpickledItems.cs

```diff
@@ -0,0 +1,134 @@
+                            castObjArr.Add((string[])arrList.ToArray(typeof(string)));
+                            break;
+                        case TypeCode.Object:
+                            Type t = ((ArrayList)arrList[0])[0].GetType();
+                            switch (Type.GetTypeCode(t))
+                            {
+                                case TypeCode.Int32:
+                                    var intArr = new List<int[]>();
+                                    foreach (ArrayList al in arrList)
+                                    {
+                                        intArr.Add((int[])al.ToArray(typeof(int)));
+                                    }
+                                    castObjArr.Add(intArr.ToArray());
+                                    break;
+                                case TypeCode.Int64:
+                                    var longArr = new List<long[]>();
+                                    foreach (ArrayList al in arrList)
+                                    {
+                                        longArr.Add((long[])al.ToArray(typeof(long)));
+                                    }
+                                    castObjArr.Add(longArr.ToArray());
+                                    break;
+                                case TypeCode.Double:
+                                    var doubleArr = new List<double[]>();
+                                    foreach (ArrayList al in arrList)
+                                    {
+                                        doubleArr.Add((double[])al.ToArray(typeof(double)));
+                                    }
+                                    castObjArr.Add(doubleArr.ToArray());
+                                    break;
+                                case TypeCode.Byte:
+                                    var byteArr = new List<byte[]>();
+                                    foreach (ArrayList al in arrList)
+                                    {
+                                        byteArr.Add((byte[])al.ToArray(typeof(byte)));
+                                    }
+                                    castObjArr.Add(byteArr.ToArray());
+                                    break;
+                                case TypeCode.String:
+                                    var stringArr = new List<string[]>();
+                                    foreach (ArrayList al in arrList)
+                                    {
+                                        stringArr.Add((string[])al.ToArray(typeof(string)));
+                                    }
+                                    castObjArr.Add(stringArr.ToArray());
+                                    break;
+                                default:
+                                    throw new NotSupportedException(
+                                        string.Format("Array of type {0} not supported yet", t));
+                            }
```

Comment:
Can you abstract this logic into a separate function and use that instead of this duplication?

Reply:
Yes, changed. Thanks!